src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -151,6 +155,53 @@ static BoundNode unwrapListElement(BoundCollectionExpression node, BoundNode ele
             }
         }
 
+        // If we have something like `List<int> l = [.. someEnumerable]`
+        // try rewrite it using `Enumerable.ToList` member if possible
+        private bool TryRewriteSingleElementSpreadToList(BoundCollectionExpression node, TypeWithAnnotations listElementType, [NotNullWhen(true)] out BoundExpression? result)
+        {
+            result = null;
+
+            if (node.Elements is not [BoundCollectionExpressionSpreadElement singleSpread])
+            {
+                return false;
+            }
+
+            var iEnumerableOfTType = _compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var iEnumerableOfElementType = iEnumerableOfTType.Construct([listElementType]);
+
+            var discardedUseSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+
+            var singleSpreadExpression = singleSpread.Expression;
+            var conversion = _compilation.Conversions.ClassifyImplicitConversionFromExpression(singleSpreadExpression, iEnumerableOfElementType, ref discardedUseSiteInfo);
```

Comment:
>ClassifyImplicitConversionFromExpression [](http://example.com/codeflow?start=54&length=40)

Using `ClassifyImplicitConversionFromExpression()` rather than `ClassifyImplicitConversionFromType()` seems surprising to me. (I think we should be converting from the *type* we would have used for enumeration rather than the original expression.)

It also looks like this set of checks is similar to the checks we're performing when optimizing adding a spread element to a list by using `List<T>.AddRange(IEnumerable<T>)` in `CreateAndPopulateList()`. And in that case, we are using `ClassifyConversionFromType()` rather than `...FromExpression()`. Would it make sense to share the code? cc @RikkiGibson

Reply:
I think yes. It seems like they are checking several of the same things.

@DoctorKrolic could you try to extract a helper based on the tryOptimizeSpreadElement used for AddRange calls. Something like `bool ShouldUseAddRangeOrToListMethod(MethodSymbol? method, BoundCollectionExpressionSpreadElement spreadElement, BoundExpression rewrittenSpreadOperand)`.